csc.exe fails with mathematical assignment operators working on nullable value types

I'm trying to upgrade a rather large codebase to VS2015. The codebase builds perfectly on VS2013.

In VS2015, csc exits with an exit code corresponding to `COR_E_FAILFAST`.

I've managed to narrow this down to mathematical operations on nullable structs with implicit conversions. This can be reproduced with the following simple case:

    using System;

    namespace NullableMathRepro
    {
        class Program
        {
            static void Main(string[] args)
            {
                int? x = 0;
                x += 5;
                Console.WriteLine("'x' is {0}", x);

                IntHolder? y = 0;
                y += 5;
                Console.WriteLine("'y' is {0}", y);
            }
        }

        struct IntHolder
        {
            private int x;

            public static implicit operator int(IntHolder ih)
            {
                return ih.x;
            }

            public static implicit operator IntHolder(int i)
            {
                return new IntHolder { x = i };
            }

            public override string ToString()
            {
                return x.ToString();
            }
        }
    }


Using mathematical assignment operators on just the `int?` works fine.
Using mathematical assignment operators on the `IntHolder` works just fine.

**Using mathematical assignment operators on the `IntHolder?` causes the compiler to crash.**

If I replace `y += 5` with `y = y.Value + 5`, it compiles and runs as expected.

Visual Studio output:

    1>------ Build started: Project: NullableMathRepro, Configuration: Debug Any CPU ------
    1>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.CSharp.Core.targets(67,5): error MSB6006: "csc.exe" exited with code -2146232797.
    ========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

If I compile and run Roslyn from master (0000c35d), this assertion fails:

https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs#L956

If I ignore that assertion, this assertion subsequently fails:
https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs#L769

(`newOwner.OriginalDefinition` is `Int32`, but `this.ContainingSymbol.OriginalDefinition` is `Nullable<T>`.)

If I ignore that subsequent assertion, it finally fails with an `InvalidCastException` on the following line ( https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs#L770 ):

    System.InvalidCastException: Unable to cast object of type 'PENamedTypeSymbolNonGeneric' to type 'Microsoft.CodeAnalysis.CSharp.Symbols.SubstitutedNamedTypeSymbol'.
       at Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol.AsMember(NamedTypeSymbol newOwner) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Symbols\MethodSymbol.cs:line 770
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.GetNullableMethod(CSharpSyntaxNode syntax, TypeSymbol nullableType, SpecialMember member) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_BinaryOperator.cs:line 1631
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.MakeLiftedUserDefinedConversionConsequence(BoundCall call, TypeSymbol resultType) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_Conversion.cs:line 957
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.RewriteLiftedUserDefinedConversion(CSharpSyntaxNode syntax, BoundExpression rewrittenOperand, MethodSymbol method, TypeSymbol rewrittenType, ConversionKind conversionKind) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_Conversion.cs:line 1019
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.RewriteUserDefinedConversion(CSharpSyntaxNode syntax, BoundExpression rewrittenOperand, MethodSymbol method, TypeSymbol rewrittenType, ConversionKind conversionKind) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_Conversion.cs:line 939
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.MakeConversion(CSharpSyntaxNode syntax, BoundExpression rewrittenOperand, Conversion conversion, TypeSymbol rewrittenType, Boolean checked, Boolean explicitCastInCode, ConstantValue constantValueOpt) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_Conversion.cs:line 557
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitCompoundAssignmentOperator(BoundCompoundAssignmentOperator node, Boolean used) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_CompoundAssignmentOperator.cs:line 65
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitUnusedExpression(BoundExpression expression) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_ExpressionStatement.cs:line 43
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionStatement(BoundExpressionStatement node) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_ExpressionStatement.cs:line 15
       at Microsoft.CodeAnalysis.CSharp.BoundExpressionStatement.Accept(BoundTreeVisitor visitor) in C:\GH\Roslyn\Binaries\Obj\CSharpCodeAnalysis\Debug\BoundNodes.xml.Generated.cs:line 2037
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 149
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitBlock(BoundBlock node) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_Block.cs:line 30
       at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor) in C:\GH\Roslyn\Binaries\Obj\CSharpCodeAnalysis\Debug\BoundNodes.xml.Generated.cs:line 1719
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 149
       at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Rewrite(CSharpCompilation compilation, MethodSymbol method, Int32 methodOrdinal, NamedTypeSymbol containingType, BoundStatement statement, TypeCompilationState compilationState, SynthesizedSubmissionFields previousSubmissionFields, Boolean allowOmissionOfConditionalCalls, DiagnosticBag diagnostics, Boolean& sawLambdas, Boolean& sawAwaitInExceptionHandler) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter.cs:line 70
       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, DiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 1151
       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 961
       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType) in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 432
       at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass20_0.<CompileNamedTypeAsTask>b__0() in C:\GH\Roslyn\src\Compilers\CSharp\Portable\Compiler\MethodCompiler.cs:line 322

--- a/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs
+++ b/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs
@@ -554,11 +554,23 @@ namespace Microsoft.CodeAnalysis.CSharp
                         @checked);
                 }
 
+                TypeSymbol userDefinedConversionRewrittenType = conversion.Method.ReturnType;
+
+                if (rewrittenOperand.Type != conversion.Method.ParameterTypes[0])
+                {
+                    Debug.Assert(rewrittenOperand.Type.IsNullableType());
+                    Debug.Assert(rewrittenOperand.Type.GetNullableUnderlyingType() == conversion.Method.ParameterTypes[0]);
+                    Debug.Assert(!userDefinedConversionRewrittenType.IsNullableType());
+
+                    // Lifted conversion, wrap return type in Nullable
+                    userDefinedConversionRewrittenType = ((NamedTypeSymbol)rewrittenOperand.Type.OriginalDefinition).Construct(userDefinedConversionRewrittenType);
+                }
+
                 BoundExpression userDefined = RewriteUserDefinedConversion(
                     syntax,
                     rewrittenOperand,
                     conversion.Method,
-                    conversion.Method.ReturnType,
+                    userDefinedConversionRewrittenType,
                     conversion.Kind);
 
                 if (userDefined.Type != conversion.BestUserDefinedConversionAnalysis.ToType)

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs
@@ -8676,5 +8676,57 @@ namespace roslynChanges
 System.Int32
 ");
         }
+
+        [Fact, WorkItem(4132, "https://github.com/dotnet/roslyn/issues/4132")]
+        public void Issue4132()
+        {
+            string source = @"
+using System;
+
+namespace NullableMathRepro
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            int? x = 0;
+            x += 5;
+            Console.WriteLine(""'x' is {0}"", x);
+
+            IntHolder? y = 0;
+            y += 5;
+            Console.WriteLine(""'y' is {0}"", y);
+        }
+    }
+
+    struct IntHolder
+    {
+        private int x;
+
+        public static implicit operator int (IntHolder ih)
+        {
+            Console.WriteLine(""operator int (IntHolder ih)"");
+            return ih.x;
+        }
+
+        public static implicit operator IntHolder(int i)
+        {
+            Console.WriteLine(""operator IntHolder(int i)"");
+            return new IntHolder { x = i };
+        }
+
+        public override string ToString()
+        {
+            return x.ToString();
+        }
+    }
+}";
+            CompileAndVerify(source: source, expectedOutput:
+@"'x' is 5
+operator IntHolder(int i)
+operator int (IntHolder ih)
+operator IntHolder(int i)
+'y' is 5");
+        }
     }
 }

--- a/src/Compilers/VisualBasic/Test/Semantic/Semantics/CompoundAssignment.vb
+++ b/src/Compilers/VisualBasic/Test/Semantic/Semantics/CompoundAssignment.vb
@@ -2704,6 +2704,65 @@ False
 ]]>)
         End Sub
 
+        <Fact, WorkItem(4132, "https://github.com/dotnet/roslyn/issues/4132")>
+        Public Sub Issue4132()
+
+            Dim compilationDef =
+<compilation>
+    <file name="a.vb">
+Imports System
+
+Namespace NullableMathRepro
+    Module Program
+        Sub Main()
+            Dim x As Integer? = 0
+            x += 5
+            Console.WriteLine("'x' is {0}", x)
+
+            Dim y As IntHolder? = 0
+            y += 5
+            Console.WriteLine("'y' is {0}", y)
+        End Sub
+    End Module
+
+    Structure IntHolder
+        Private x As Integer
+
+        Public Shared Widening Operator CType(ih As IntHolder) As Integer
+            Console.WriteLine("operator int (IntHolder ih)")
+            Return ih.x
+        End Operator
+
+        Public Shared Widening Operator CType(i As Integer) As IntHolder
+            Console.WriteLine("operator IntHolder(int i)")
+            Return New IntHolder With {.x = i}
+        End Operator
+
+        Public Shared Operator +(x As IntHolder, y As Integer) As Integer
+            Return CInt(x) + y
+        End Operator
+        Public Overrides Function ToString() As String
+            Return x.ToString()
+        End Function
+
+    End Structure
+End Namespace
+    </file>
+</compilation>
+
+            Dim compilation = CompilationUtils.CreateCompilationWithMscorlibAndVBRuntime(compilationDef, TestOptions.ReleaseExe)
+
+            CompileAndVerify(compilation,
+            <![CDATA[
+'x' is 5
+operator IntHolder(int i)
+operator int (IntHolder ih)
+operator IntHolder(int i)
+'y' is 5
+]]>)
+        End Sub
+
     End Class
 
 End Namespace
+

